Add Readiness and liveliness probes to nfd-master

ArangoGutierrez opened the pull request:
Fixes #522 

This patch seeks to:

- https://github.com/kubernetes-sigs/node-feature-discovery/commit/273cfffe62936300f575bca72205bff0cc8bae3d enable readinessprobe for nfd-master
- https://github.com/kubernetes-sigs/node-feature-discovery/pull/563/commits/f4e8e20ea5c963c0cb4610f702abf0ed889b90a9 add livenessprobe for nfd-master

ArangoGutierrez commented:
/cc @MaxRink

ArangoGutierrez commented:
/assign @marquiz

zvonkok commented:
Hmm, why are we introducing a new HTTP server when we can do grpc liveness probes? https://kubernetes.io/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/

Ethyling commented on pkg/nfd-master/nfd-master.go:
```diff
@@ -195,6 +197,21 @@ func (m *nfdMaster) Run() error {
+	err = r.Set()
+	if err != nil {
+		klog.Error(err, "")
+		os.Exit(1)
```
Do we really want to exit instead of returning the error?

zvonkok commented on nfd-daemonset-combined.yaml.template:
```diff
@@ -69,6 +69,22 @@ spec:
                 fieldPath: spec.nodeName
           image: gcr.io/k8s-staging-nfd/node-feature-discovery:master
           imagePullPolicy: Always
+<<<<<<< HEAD
```
Merge not done?

ArangoGutierrez replied:
Rebase not completed :(

ArangoGutierrez replied:
Done now thanks for the catch

ArangoGutierrez commented:
@zvonkok PTAL , thanks for the previous reviews

marquiz commented on pkg/ready/ready.go:
```diff
@@ -0,0 +1,70 @@
+		}
+		return err
+	}
+	klog.Infof("Readiness Probe set to %s", r.FileName)
```
Should this be debug (i.e. `V(1)`), wdyt?

ArangoGutierrez replied:
done

marquiz commented on pkg/ready/ready.go:
```diff
@@ -0,0 +1,70 @@
+func (r fileReady) Set() error {
+	f, err := os.Create(r.FileName)
+	if err != nil {
+		if os.IsExist(err) {
```
Should we log this, too? WDYT?

ArangoGutierrez replied:
done

marquiz reviewed (commented):
Thanks @ArangoGutierrez for this, looks good! Just few comments/questions

ArangoGutierrez commented on nfd-master.yaml.template:
```diff
@@ -102,6 +115,13 @@ spec:
             runAsNonRoot: true
           command:
             - "nfd-master"
+## Enable Readiness probe
+          volumeMounts:
+          - mountPath: /tmp
+            name: tmp
+      volumes:
+      - name: tmp
+        emptyDir: {}
```
is a READ only container, only volumes are writable

marquiz replied:
Ach, yes of course 🤦

marquiz reviewed (commented):
Does the readiness probe actually bring any value to us, or just complication and more LOC? We set the "flag" at startup and unset on os.Exit but no toggling during runtime. Thus, I think liveness probe covers everything and the readiness probe could/should be dropped. WDYT?

ArangoGutierrez commented on pkg/nfd-master/nfd-master.go:
```diff
@@ -205,10 +225,18 @@ func (m *nfdMaster) Run() error {
 			klog.Infof("shutting down nfd-master")
+			if err = r.Unset(); err != nil {
+				klog.Error(err, "")
+				os.Exit(1)
```
done

ArangoGutierrez commented on pkg/nfd-master/nfd-master.go:
```diff
@@ -205,10 +225,18 @@ func (m *nfdMaster) Run() error {
 		case <-grpcErr:
+			if err = r.Unset(); err != nil {
+				klog.Error(err, "")
+				os.Exit(1)
```
done

ArangoGutierrez commented on nfd-daemonset-combined.yaml.template:
```diff
@@ -78,13 +91,22 @@ spec:
+            exec:
+              command: ["/usr/bin/grpc_health_probe", "-addr=:8080"]
+            initialDelaySeconds: 10
+            periodSeconds: 10
```
good catch

ArangoGutierrez commented on Dockerfile:
```diff
@@ -21,6 +21,10 @@ RUN make install VERSION=$VERSION HOSTMOUNT_PREFIX=$HOSTMOUNT_PREFIX
+# Download the grpc_health_probe bin
+RUN GRPC_HEALTH_PROBE_VERSION=v0.3.1 && \
+    wget -qO/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-amd64 && \
+    chmod +x /bin/grpc_health_probe
```
done

ArangoGutierrez commented:
> Does the readiness probe actually bring any value to us, or just complication and more LOC? We set the "flag" at startup and unset on os.Exit but no toggling during runtime. Thus, I think liveness probe covers everything and the readiness probe could/should be dropped. WDYT?

Well, by definition the `ReadinessProbe` just tells Kubernetes that the service is ready to start getting traffic, that is why the ready func is after a successful `Run gRPC server` run, and we remove the file from the volume when the service is being shut down, in case the volume will be re-used for example in a rolling-update event. I think is serves it's purpose as it is. 

Worth mentioning this is just for the Master (that is the service that gets traffic) I didn't added this checks to the worker bc it can be used in a one-shot run, so is no worth adding long run checks to it.

(Addressed all your comments 😃 )

marquiz commented:
> Well, by definition the `ReadinessProbe` just tells Kubernetes that the service is ready to start getting traffic, that is why the ready func is after a successful `Run gRPC server` run, and we remove the file from the volume when the service is being shut down, in case the volume will be re-used for example in a rolling-update event. I think is serves it's purpose as it is.

My thinking was that we virtually have no time when the pod is running but it's not "Ready". The startup time of nfd-master is very fast and we never toggle the readiness on/off at runtime, just turn if of at pod termination time. I'm not sure if the added LOC (with maintenance burden( brings any added value. If we want to have a readiness probe, maybe se should just re-use the grpc probe because that **really** tells if we're ready to accept traffic. WDYT?

> Worth mentioning this is just for the Master (that is the service that gets traffic) I didn't added this checks to the worker bc it can be used in a one-shot run, so is no worth adding long run checks to it.

Yeah, I don't think we need to try to add any checks there for now.

ArangoGutierrez commented:
> > Well, by definition the `ReadinessProbe` just tells Kubernetes that the service is ready to start getting traffic, that is why the ready func is after a successful `Run gRPC server` run, and we remove the file from the volume when the service is being shut down, in case the volume will be re-used for example in a rolling-update event. I think is serves it's purpose as it is.
> 
> My thinking was that we virtually have no time when the pod is running but it's not "Ready". The startup time of nfd-master is very fast and we never toggle the readiness on/off at runtime, just turn if of at pod termination time. I'm not sure if the added LOC (with maintenance burden( brings any added value. If we want to have a readiness probe, maybe se should just re-use the grpc probe because that **really** tells if we're ready to accept traffic. WDYT?
> 
> > Worth mentioning this is just for the Master (that is the service that gets traffic) I didn't added this checks to the worker bc it can be used in a one-shot run, so is no worth adding long run checks to it.
> 
> Yeah, I don't think we need to try to add any checks there for now.

Readiness also via Grpc, I agree, working on the fix

marquiz reviewed (changes requested):
Looks good and simple now 👍 

Just one thing: I think we could drop ` pkg/ready/ready.go` now.

ArangoGutierrez commented:
> Looks good and simple now +1
> 
> Just one thing: I think we could drop ` pkg/ready/ready.go` now.

Done

marquiz reviewed (approved):
This is a nice improvement @ArangoGutierrez!
/lgtm

dbaker-rh commented:
FYI - this probe addition has broken installs that make use of TLS (grpc_health_check also needs to use tls if the listener is).  I offer #710 to address this in the kustomize framework, and will probably bash together a patch for the helm chart shortly to do the same.
